src/Microsoft.DotNet.VersionTools/Automation/PullRequestCreator.cs

```diff
@@ -0,0 +1,205 @@
+using System.Threading.Tasks;
+
+namespace Microsoft.DotNet.VersionTools.Automation
+{
+    public class PullRequestCreator
+    {
+        private GitHubAuth _auth;
+
+        public string BranchPrefix { get; set; } = "UpdateDependencies";
+
+        public bool AlwaysCreateNewPullRequest { get; set; }
+
+        public GitHubBranch UpstreamBranch { get; }
+        public GitHubProject Origin { get; }
+
+        public string GitAuthorName { get; }
+
+        public PullRequestCreator(
+            GitHubAuth auth,
+            string projectName,
+            string upstreamOwner,
+            string upstreamBranch,
+            string gitAuthorName)
+            : this(
+                auth,
+                new GitHubProject(projectName, auth.User),
+                new GitHubBranch(upstreamBranch, new GitHubProject(projectName, upstreamOwner)),
+                gitAuthorName)
+        {
+        }
+
+        public PullRequestCreator(
+            GitHubAuth auth,
+            GitHubProject origin,
+            GitHubBranch upstreamBranch,
+            string gitAuthorName = null)
+        {
+            _auth = auth;
+
+            Origin = origin;
+            UpstreamBranch = upstreamBranch;
+
+            GitAuthorName = gitAuthorName ?? auth.User;
+        }
+
+        public async Task CreateAsync(
+            string commitMessage,
+            string title,
+            string description)
+        {
+            using (var client = new GitHubClient(_auth))
+            {
+                var originBranch = new GitHubBranch(
+                    $"{BranchPrefix}{DateTime.UtcNow.ToString("yyyyMMddhhmmss")}",
+                    Origin);
+
+                GitHubPullRequest pullRequestToUpdate = null;
+
+                if (!AlwaysCreateNewPullRequest)
+                {
+                    pullRequestToUpdate = await client.FirstPullRequestWithHeadOrDefaultAsync(
+                        Upstream,
+                        BranchPrefix,
+                        _auth.User);
+
+                    if (pullRequestToUpdate == null)
+                    {
+                        Trace.TraceInformation($"No existing pull request found.");
+                    }
+                    else
+                    {
+                        Trace.TraceInformation(
+                            $"Pull request already exists for {BranchPrefix} in {Upstream.Segments}. " +
+                            $"#{pullRequestToUpdate.Number}, '{pullRequestToUpdate.Title}'");
+
+                        if (CanUpdatePullRequest(client, pullRequestToUpdate))
+                        {
+                            originBranch = new GitHubBranch(
+                                pullRequestToUpdate.Head.Ref,
+                                originBranch.Project);
+                        }
+                        else
+                        {
+                            Trace.TraceWarning($"Can't update pull request #{pullRequestToUpdate.Number}.");
```

Comment:
Hmm, well this is expected to happen once in a while, but true that if it happens without someone expecting it then that would be a problem. We do need some way of telling when this happens. I'm not sure if opening a new PR is the right way, though, since if this was intentional, we don't want someone else to see a newer green PR and merge it, making it harder to apply whatever fixup is being done to the old PR. Maybe it could just append to the description (I already have the API for that) or add a comment?

---
In reply to: [72437681](https://github.com/dotnet/buildtools/pull/919#discussion_r72437681) [](ancestors = 72437681)